net/graphsync_fetcher_test.go

```diff
@@ -43,58 +43,78 @@ func TestGraphsyncFetcher(t *testing.T) {
 	bv := th.NewFakeBlockValidator()
 	pid0 := th.RequireIntPeerID(t, 0)
 	builder := chain.NewBuilder(t, address.Undef)
+	keys := types.MustGenerateKeyInfo(1, 42)
+	mm := types.NewMessageMaker(t, keys)
+	alice := mm.Addresses()[0]
 
 	ssb := selectorbuilder.NewSelectorSpecBuilder(ipldfree.NodeBuilder())
 	layer1Selector, err := ssb.ExploreFields(func(efsb selectorbuilder.ExploreFieldsSpecBuilder) {
 		efsb.Insert("messages", ssb.Matcher())
 		efsb.Insert("messageReceipts", ssb.Matcher())
 	}).Selector()
 	require.NoError(t, err)
-	gsSelector, err := ssb.ExploreRecursive(1, ssb.ExploreFields(func(efsb selectorbuilder.ExploreFieldsSpecBuilder) {
-		efsb.Insert("messages", ssb.Matcher())
-		efsb.Insert("messageReceipts", ssb.Matcher())
-		efsb.Insert("parents", ssb.ExploreUnion(
-			ssb.ExploreAll(ssb.Matcher()),
-			ssb.ExploreIndex(0, ssb.ExploreRecursiveEdge()),
-		))
-	})).Selector()
-	require.NoError(t, err)
-	gsSelectorRound2, err := ssb.ExploreRecursive(4, ssb.ExploreFields(func(efsb selectorbuilder.ExploreFieldsSpecBuilder) {
-		efsb.Insert("messages", ssb.Matcher())
-		efsb.Insert("messageReceipts", ssb.Matcher())
-		efsb.Insert("parents", ssb.ExploreUnion(
-			ssb.ExploreAll(ssb.Matcher()),
-			ssb.ExploreIndex(0, ssb.ExploreRecursiveEdge()),
-		))
-	})).Selector()
-	require.NoError(t, err)
+	recursiveSelector := func(levels int) selector.Selector {
+		s, err := ssb.ExploreRecursive(levels, ssb.ExploreFields(func(efsb selectorbuilder.ExploreFieldsSpecBuilder) {
+			efsb.Insert("parents", ssb.ExploreUnion(
+				ssb.ExploreAll(
+					ssb.ExploreFields(func(efsb selectorbuilder.ExploreFieldsSpecBuilder) {
+						efsb.Insert("messages", ssb.Matcher())
+						efsb.Insert("messageReceipts", ssb.Matcher())
+					}),
+				),
+				ssb.ExploreIndex(0, ssb.ExploreRecursiveEdge()),
+			))
+		})).Selector()
+		require.NoError(t, err)
+		return s
+	}
 	pid1 := th.RequireIntPeerID(t, 1)
 	pid2 := th.RequireIntPeerID(t, 2)
 
+	// Returns an array of IPLD nodes for the headers, messages, and receipts of some blocks.
+	ipldBlocks := func(bs ...*types.Block) []format.Node {
+		nodes := make([]format.Node, 3*len(bs))
+		for i, b := range bs {
+			nodes[3*i] = b.ToNode()
+			m, err := builder.LoadMessages(ctx, b.Messages)
+			require.NoError(t, err)
+			nodes[3*i+1] = types.MessageCollection(m).ToNode()
+			r, err := builder.LoadReceipts(ctx, b.MessageReceipts)
+			require.NoError(t, err)
+			nodes[3*i+2] = types.ReceiptCollection(r).ToNode()
+		}
+		return nodes
+	}
```

Comment:
Most of the change in this file is to return the correct data (i.e. including message/receipt lists) to the fetcher, which now checks for it.